include/picongpu/fields/MaxwellSolver/ArbitraryOrderFDTD/Derivative.hpp

```diff
@@ -0,0 +1,312 @@
+         * @param data position in the data box to compute derivative at
+         */
+        template< typename T_DataBox >
+        HDINLINE typename T_DataBox::ValueType operator()(
+            T_DataBox const & data ) const
+        {
+            // Define shorthand type to access DataBox
+            using Index = pmacc::DataSpace< simDim >;
+
+            // Define indice vectors for data access
+            auto lowerIndex = Index{}; // Vector initialized with zeros
+            
+            auto upperIndex = pmacc::math::basisVector<
+                Index,
+                T_direction
+            >();
+
+            auto const indexIncrement = pmacc::math::basisVector<
+                Index,
+                T_direction
+            >();
+
+            AOFDTDWeights< T_neighbors > const weights{};
+
+            // shortest distance finite difference as initial value
+            auto finiteDifference = weights[0] *
+                ( data( upperIndex ) - data( lowerIndex ) );
+
+            // Compute next finite differences according to order
+            for ( uint32_t l = 1; l < T_neighbors; ++l )
+            {
+                lowerIndex -= indexIncrement;
+                upperIndex += indexIncrement;
+
+                finiteDifference += weights[l] *
+                ( data( upperIndex ) - data( lowerIndex ) );
+            }
+
+            return finiteDifference / cellSize[ T_direction ];
+        }
+    };
+
+
+    /** Functor for backward derivative along the given direction used in ArbitraryOrderFDTD solver
+     * 
+     * Compute an approximation of the derivative of a field f by a finite difference of
+     * order 2 * T_neighbors, where T_neighbors is the number of neighbors
+     * used to calculate the finite difference.
+     * 
+     * This finite difference approximation for the derivative is computed on a staggered grid,
+     * i.e. the derivative will be known at a position i-1/2, if the field f is known 
+     * at 2 * T_neighbors grid nodes i - T_neighbors, i - T_neighbors + 1, ..., i + T_neighbors - 1.
+     
+     * The finite difference calculation can be expressed as a sum of finite differences where the 
+     * distance of field components used in individual finite differences computations increases
+     *     D_x f(i-1/2) = sum_{l=0}^{T_neighbors-1} g_l^{2T_neighbors} * ( f(i+l) - f(i-1-l) ) / dx,
+     * where D_x is the derivative operator along x, dx the grid spacing along x, and 
+     * g_l^{2T_neighbors} weightings for the finite differences of different distance l from the point i
+     * of computation.
+     * 
+     * @tparam T_neighbors Number of neighbors used to calculate
+     *                     the derivative from finite differences.
+     *                     Order of derivative approximation is
+     *                     2 * T_neighbors
+     * 
+     * @tparam T_direction direction to take derivative in, 0 = x, 1 = y, 2 = z
+     */
+    template<
```

Comment:
I suggest to avoid copy past of the full class by using a helper class.

```C++
namespace detail
{    
    template<
        uint32_t T_lowerMargin,
        uint32_t T_upperMargin,
        uint32_t T_neighbors,
        uint32_t T_direction
    >
    struct DerivativeFunctor
    {
        //! Lower margin
        using LowerMargin = typename pmacc::math::CT::mul<
            typename pmacc::math::CT::make_Int<
                simDim,
                T_neighbors
            >::type,
            typename pmacc::math::CT::make_BasisVector<
                simDim,
                T_direction,
                int
            >::type
        >::type;

        //! Upper margin
        using UpperMargin = typename pmacc::math::CT::mul<
            typename pmacc::math::CT::make_Int<
                simDim,
                T_neighbors - 1
            >::type,
            typename pmacc::math::CT::make_BasisVector<
                simDim,
                T_direction,
                int
            >::type
        >::type;

        /** Return derivative value at the given point
         *
         * @tparam T_DataBox data box type with field data
         * @param data position in the data box to compute derivative at
         */
        template< typename T_DataBox >
        HDINLINE typename T_DataBox::ValueType operator()(
            T_DataBox const & data ) const
        {
            // Define shorthand type to access DataBox
            using Index = pmacc::DataSpace< simDim >;

            // Define indice vectors for data access
            auto lowerIndex = -pmacc::math::basisVector<
                Index,
                T_direction
            >(); // Vector initialized with zeros

            auto upperIndex = Index{};

            auto const indexIncrement = pmacc::math::basisVector<
                Index,
                T_direction
            >();

            AOFDTDWeights< T_neighbors > const weights{};

            // shortest distance finite difference as initial value
            auto finiteDifference = weights[0] *
                ( data( upperIndex ) - data( lowerIndex ) );

            // Compute next finite differences according to order
            for ( uint32_t l = 1; l < T_neighbors; ++l )
            {
                lowerIndex -= indexIncrement;
                upperIndex += indexIncrement;

                finiteDifference += weights[l] *
                ( data( upperIndex ) - data( lowerIndex ) );
            }

            return finiteDifference / cellSize[ T_direction ];
        }
    };
    
} // namespace detail

    template<
        uint32_t T_neighbors,
        uint32_t T_direction
    >
    using ForwardDerivativeFunctor = detail::DerivativeFunctor<T_neighbors -1, T_neighbors, T_neighbors, T_direction>;
    
    template<
        uint32_t T_neighbors,
        uint32_t T_direction
    >
    using BackwardDerivativeFunctor = detail::DerivativeFunctor<T_neighbors, T_neighbors - 1, T_neighbors, T_direction>;
```